src/Core/NETPortable/NetPortableProfileTable.cs

```diff
@@ -14,95 +13,95 @@ public static class NetPortableProfileTable
+            }
+        }
 
         public static NetPortableProfile GetProfile(string profileName)
         {
-            if (String.IsNullOrEmpty(profileName))
+            if (string.IsNullOrEmpty(profileName))
             {
                 throw new ArgumentException(CommonResources.Argument_Cannot_Be_Null_Or_Empty, "profileName");
             }
 
             // Original behavior fully preserved, as we first try the original behavior.
             // NOTE: this could be a single TryGetValue if this collection was kept as a dictionary...
-            if (Profiles.Contains(profileName))
+            if (Compiled.Profiles.Contains(profileName))
             {
-                return Profiles[profileName];
+                return Compiled.Profiles[profileName];
             }
 
             // If we didn't get a profile by the simple profile name, try now with 
             // the custom profile string (i.e. "net40-client")
             NetPortableProfile result = null;
-            _portableProfilesByCustomProfileString.TryGetValue(profileName, out result);
+            Compiled.PortableProfilesByCustomProfileString.TryGetValue(profileName, out result);
 
             return result;
         }
 
         internal static NetPortableProfileCollection Profiles
         {
-            get
+            set
             {
-                if (_portableProfiles == null)
+                if (value == null)
+                {
+                    _portableProfiles = null;
+                }
+                else
                 {
-                    // We use the setter so that we can consistently set both the 
-                    // existing collection as well as the CustomProfileString-indexed one.
-                    // This keeps both in sync.
-                    Profiles = BuildPortableProfileCollection();
-                    CreateOptionalFrameworksDictionary();
+                    _portableProfiles = new CompiledNetPortableProfileCollection(value);
                 }
 
-                return _portableProfiles;
-            }
-            set
-            {
-                // This setter is only for Unit Tests.
-                _portableProfiles = value;
-                _portableProfilesByCustomProfileString = _portableProfiles.ToDictionary(x => x.CustomProfileString);
-                CreateOptionalFrameworksDictionary();
+                // This setter is only for unit tests and is NOT thread-safe.
+                // Reset the lazily loaded profiles.
+                _lazyPortableProfiles = new Lazy<CompiledNetPortableProfileCollection>(() => new CompiledNetPortableProfileCollection(BuildPortableProfileCollection()));
             }
         }
 
-        private static void CreateOptionalFrameworksDictionary()
-        {            
-            if (_portableProfiles != null)
+        private static IDictionary<string, NetPortableProfile> CreatePortableProfilesByCustomProfileString(NetPortableProfileCollection profileCollection)
```

Comment:
The method solely is consumed by `CompiledNetPortableProfileCollection` constructor. Does it make sense to move it there then?